Improve fee estimation

morcos opened the pull request:
This PR is partly a refactor, partly bug fix, and partly improved robustness in edge cases.  Outside of the edge cases (when your node is behind) the results are unchanged.
 
Explanations of the changes in the individual commits.

morcos committed:
Remove extraneous LogPrint from fee estimation

Once priority estimation was removed, not all transactions in the mempool are tracked in the fee estimation mempool tracking.  So there is no error if a transaction is not found for removal.

morcos committed:
Don't track transactions at all during IBD.

This was an oversight, where blocks and mempool tracking were ignored during IBD, but transactions that arrived during IBD but were included in blocks after IBD were not ignored.

morcos committed:
Remove member variable hadNoDependencies from CTxMemPoolEntry

Fee estimation can just check its own mapMemPoolTxs to determine the same information.  Note that now fee estimation for block processing must happen before those transactions are removed, but this shoudl be a speedup.

morcos committed:
rename bool to validFeeEstimate

morcos committed:
Always update fee estimates on new blocks.

All decisions about whether the transactions are valid data points are made at the time the transaction arrives. Updating on blocks all the time will now cause stale fee estimates to decay quickly when we restart a node.

morcos committed:
Add IsCurrentForFeeEstimatation

Make a more conservative notion of whether the node is caught up to the rest of the network and only count transactions as fee estimation data points if the node is caught up.

morcos committed:
Add extra logging to processBlock in fee estimation.

morcos committed:
Pass pointers to existing CTxMemPoolEntries to fee estimation

rebroad commented on src/main.cpp:
```diff
@@ -1173,6 +1173,17 @@ std::string FormatStateMessage(const CValidationState &state)
+        return false;
+    if (chainActive.Tip()->GetBlockTime() < (GetTime() - MAX_FEE_ESTIMATION_TIP_AGE))
+        return false;
+    if (chainActive.Height() < pindexBestHeader->nHeight - 1)
```
Why -1?

rebroad commented on src/main.cpp:
```diff
@@ -1173,6 +1173,17 @@ std::string FormatStateMessage(const CValidationState &state)
+        return false;
+    if (chainActive.Height() < pindexBestHeader->nHeight - 1)
+        return false;
+    return true;
```
nit - usually a blank line before this line (but probably not a standard)

rebroad commented on src/policy/fees.cpp:
```diff
@@ -343,34 +337,33 @@ void CBlockPolicyEstimator::processTransaction(const CTxMemPoolEntry& entry, boo
     // blocksToConfirm is 1-based, so a transaction included in the earliest
     // possible block has confirmation count of 1
-    int blocksToConfirm = nBlockHeight - entry.GetHeight();
+    int blocksToConfirm = nBlockHeight - entry->GetHeight();
```
what did this fix please?

morcos replied:
Please see commit message "Pass pointers to existing CTxMemPoolEntries to fee estimation"
The idea is this would save unnecessary copying.

morcos commented:
rebased

gmaxwell commented:
Concept ACK.

jonasschnelli commented:
Ran this code on top of master on one of my nodes.
Just compared it to the node running master (same network but independent mempool), see below.
The differences for a conf. target of 2 blocks are quite large.

Next step would be to run master and this PR with the same mempool and maybe dump 2,4,8,20 target every ~10mins.

```
masternode ~/node/bitcoin $ ./src/bitcoin-cli getmempoolinfo
{
  "size": 22022,
  "bytes": 133099706,
  "usage": 252978016,
  "maxmempool": 300000000,
  "mempoolminfee": 0.00001014
}
masternode ~/node/bitcoin $ ./src/bitcoin-cli estimatesmartfee 2
{
  "feerate": 0.00761447,
  "blocks": 2
}
masternode ~/node/bitcoin $ ./src/bitcoin-cli estimatesmartfee 3
{
  "feerate": 0.00069881,
  "blocks": 3
}
masternode ~/node/bitcoin $ ./src/bitcoin-cli estimatesmartfee 8
{
  "feerate": 0.00069881,
  "blocks": 8
}

```

```
thisprnode ~/bitcoin/bitcoin $ ./src/bitcoin-cli getmempoolinfo
{
  "size": 24593,
  "bytes": 140534928,
  "usage": 268029424,
  "maxmempool": 300000000,
  "mempoolminfee": 0.00001000
}
thisprnode ~/bitcoin/bitcoin $ ./src/bitcoin-cli estimatesmartfee 2
{
  "feerate": 0.00083575,
  "blocks": 2
}
thisprnode ~/bitcoin/bitcoin $ ./src/bitcoin-cli estimatesmartfee 3
{
  "feerate": 0.00077770,
  "blocks": 3
}
thisprnode ~/bitcoin/bitcoin $ ./src/bitcoin-cli estimatesmartfee 8
{
  "feerate": 0.00069872,
  "blocks": 8
}
```

morcos commented:
@jonasschnelli I simulated this PR's code against master and the differences were very slight.  If your testing setups were identical (started with the same fee estimates and started the nodes at the same time), then the way you might see substantial differences is if your testing nodes have to catch up a little bit when you start them, then this code will decay your old estimates and the old code wouldn't.  So this code will more quickly reflect current conditions instead of still having a heavy weight for old fee estimates.  Furthermore if you started this code after being down for more than a few weeks, you might even start seeing messages about fee estimates need to warm up as the historical data will have decayed so much.

sdaftuar commented on src/policy/fees.cpp:
```diff
@@ -392,6 +382,9 @@ void CBlockPolicyEstimator::processBlock(unsigned int nBlockHeight,
     for (unsigned int i = 0; i < entries.size(); i++)
         processBlockTx(nBlockHeight, entries[i]);
 
+    // Update best seen height after removals so they find the right mempool txs
```
Can you clarify this comment?  I have been staring at this for a while and can't figure out what you meant.

sdaftuar commented on src/main.h:
```diff
@@ -128,6 +128,8 @@ static const int64_t BLOCK_DOWNLOAD_TIMEOUT_PER_PEER = 500000;
 static const unsigned int DEFAULT_LIMITFREERELAY = 0;
 static const bool DEFAULT_RELAYPRIORITY = true;
 static const int64_t DEFAULT_MAX_TIP_AGE = 24 * 60 * 60;
+/** Maximum age of our tip for us to be considered current for fee estimation */
```
nit: Can you update the comment to include the unit (ie seconds)?

sdaftuar commented:
Code review ACK.  Will test.

sdaftuar commented on src/policy/fees.cpp:
```diff
@@ -343,10 +336,8 @@ void CBlockPolicyEstimator::processTransaction(const CTxMemPoolEntry& entry, boo
-        // This transaction depended on other transactions in the mempool to
-        // be included in a block before it was able to be included, so
-        // we shouldn't include it in our calculations
+    if (!removeTx(entry.GetTx().GetHash())) {
```
Does calling this here result in one extra call to `removeTx()` for each in-mempool, in-block transaction?  It looks like we also will call this in `CTxMempool::removeUnchecked`, which I guess is harmless, but worth commenting.

morcos committed:
Add clarifying comments to fee estimation

morcos committed:
Fix edge case with stale fee estimates

morcos commented:
- Addressed @sdaftuar's comment requests.
- Fixed a pre-existing rare edge case which would print errors and make estimates slightly off.
- Fixed a bug in this PR.

@gmaxwell once you see the BUGFIX commit, I'll squash just that one into the right place in the history.

sdaftuar commented on src/main.cpp:
```diff
@@ -1173,6 +1173,17 @@ std::string FormatStateMessage(const CValidationState &state)
+{
+    if (IsInitialBlockDownload())
+        return false;
+    if (chainActive.Tip()->GetBlockTime() < (GetTime() - MAX_FEE_ESTIMATION_TIP_AGE))
```
nit: I think we should `AssertLockHeld(cs_main)` here.

sdaftuar commented:
Tested this code in simulation on 10 days in July, comparing the transactions tracked and the fee estimates returned before and after this PR.  ACK apart from the AssertLockHeld nit above.

morcos commented:
dancing again on the grave of main.cpp 

This was a clean rebase and I added @sdaftuar's AssertLockHeld.  

In a couple days I'll squash the last two commits...

morcos commented:
rebased (and squashed in the bugfix and assert)

@fanquake please milestone for 0.14.0

sdaftuar commented:
re-ACK 98fffcb

morcos commented:
Simple rebase due to expected conflict on function signature of CTxMemPool:removeForBlock

sdaftuar commented:
re-ACK 761a9be

sipa commented on src/policy/fees.cpp:
```diff
@@ -281,19 +281,17 @@ void TxConfirmStats::removeTx(unsigned int entryHeight, unsigned int nBestSeenHe
+        mapMemPoolTxs.erase(hash);
+        return true;
+    }
+    else {
```
Coding style nit: `else` on the same line.

sipa commented on src/txmempool.cpp:
```diff
@@ -594,14 +594,14 @@ void CTxMemPool::removeConflicts(const CTransaction &tx)
         indexed_transaction_set::iterator i = mapTx.find(hash);
         if (i != mapTx.end())
-            entries.push_back(*i);
+            entries.push_back(&(*i));
```
Parentheses aren't needed here.

sipa reviewed (commented):
Code review ACK. I don't understand the fee estimation code enough to judge the changes to the semantics.

morcos commented:
Rebased to fix a silent merge conflict (the change in wallet.cpp)
Also fixed @sipa's nits, otherwise unchanged from reviewed code.

gmaxwell commented:
reviewed and tested this before it got conflicted ACK.  needs rebase.

morcos commented:
rebased

sdaftuar commented:
re-ACK 44b64b933dec10f5ffcd7f34e7bf5e4ed97f671e

sipa commented:
utACK 44b64b933dec10f5ffcd7f34e7bf5e4ed97f671e
